### ISSUE TITLE
Model does not learn anything during fine-tuning with Flash Attention

### ISSUE THREAD

**Issue by hsa-online** [ai-bond/flash-attention-v100#44]
Hello!

Before my experiments I ran the flash-attention-v100 tests and all of them passed.

But I have a strange issue on fine-tuning of Qwen2.5-0.5B (dataset is relatively small and contains about 9K examples).
Without Flash Attention the fine-tuned model gives better accuracy (on test data) than the base one, 
but with Flash Attention I see strange LR change and the model not learns.

I run fine-tuning using the llamafactory-cli:

```
llamafactory-cli train 
  --stage sft \
  --finetuning_type full \
  --fp16 \
  --learning_rate 5e-6 \
  --lr_scheduler_type cosine \
  --warmup_ratio 0.03 \
  --num_train_epochs 1.0 \
  --optim adamw_torch \
  ... then input, output, etc. etc.
```

On training I see LR values like these ones:
```
3.295e-07 
4.764e-07 
3.081e-07 
9.588-08 
7.634e-12
```

Then I remove all results, use the same base checkpoint of Qwen and enable FA by just adding one row to the command line above:
```
 --flash_attn fa2 \
```

Nothing more is changed in the input.

But this changes LR values and model does not learn more:
```
2.745e-07 
4.982e-07 
4.161e-07 
2.461e-07 
8.243e-08
```

**Comment by ai-bond** [comment]
Можешь и на русском ) Если не стесняешься ) Можешь немного инфы дополнительной? Чтобы делать SFT нужно понимать какую модель ты дообучаешь и как. В репе сейчас нет, например RotaryEmbedding, cross_entropy и прочего ) Для SFT одного FA мало )

**Comment by hsa-online** [comment]
Хмм... я не вполне понял вопрос. 

Модель "Qwen2.5-0.5B Base" (https://huggingface.co/Qwen/Qwen2.5-0.5B), а настройки я вроде все уже скинул.

Ну, за исключением

```
  --template default \
  --per_device_train_batch_size $BATCH \
  --gradient_accumulation_steps $G_STEPS \
  --model-name-or-path $MODEL_DIR \
  --logging_steps $LOGGING_STEPS \
  --save_steps $SAVE_STEPS \
  --output-dir $OUTPUT_DIR \
  --dataset-dir . \

```
Но от них результат не зависит.
Ничего своего я никуда не добавлял: FA из этого репозитория, всё остальное через pip install.

**Comment by ai-bond** [comment]
Try plz with new one merged pull #45 or just with last src code.
Changes was made not special for u but think in your case also must work. 

This 06b7189f2b16b64622d91688ba0a65bc77322143 and ec6308598d94c23d87250a278428b9b5e251b438 and 70e1c978995618726ae58a20b62b8791f56cb200 commits are fixes edge mma computation cases that's are close to "attention behaviour"

Let me know train results plz.
Thnx.

**Comment by ai-bond** [comment]
> `I have a strange issue on fine-tuning of Qwen2.5-0.5B`

```
[INFO|2026-09-22 19:57:48] modeling_utils.py:769 >> loading weights file model.safetensors from cache at .cache/huggingface/hub/models--Qwen--Qwen2.5-.5B/model.safetensors
[INFO|2026-09-22 19:57:48] modeling_utils.py:839 >> Will use dtype=torch.bfloat16 as defined in model's config object
```

> ` --flash_attn fa2`
```
[INFO|2026-09-22 19:57:49] dynamic_module_utils.py:441 >> Could not locate the custom_generate/generate.py inside Qwen/Qwen2.5-0.5B.
[INFO|2026-09-22 19:57:49] logging.py:144 >> Gradient checkpointing enabled.
[INFO|2026-09-22 19:57:49] logging.py:144 >> Using FlashAttention-2 for faster training and inference.
[INFO|2026-09-22 19:57:49] logging.py:144 >> Upcasting trainable params to float32.
[INFO|2026-09-22 19:57:49] logging.py:144 >> Fine-tuning method: Full
[INFO|2026-09-22 19:57:49] logging.py:144 >> trainable params: 494,032,768 || all params: 494,032,768 || trainable%: 100.0000
[INFO|2026-09-22 19:57:53] trainer.py:1471 >>   Num Epochs = 4
[INFO|2026-09-22 19:57:53] trainer.py:1472 >>   Num update steps per epoch = 125
[INFO|2026-09-22 19:57:53] trainer.py:1473 >>   Instantaneous batch size per device = 1
[INFO|2026-09-22 19:57:53] trainer.py:1476 >>   Total train batch size (w. parallel, distributed & accumulation) = 8
[INFO|2026-09-22 19:57:53] trainer.py:1477 >>   Gradient Accumulation steps = 4
[INFO|2026-09-22 19:57:53] trainer.py:1478 >>   Total optimization steps = 500
[INFO|2026-09-22 19:57:53] trainer.py:1479 >>   Number of trainable parameters = 494,032,768
[WARNING|2026-09-22 19:57:54] logging.py:340 >> Casting fp32 inputs back to torch.float16 for flash-attn compatibility.
[WARNING|2026-09-22 19:57:54] logging.py:340 >> Casting fp32 inputs back to torch.float16 for flash-attn compatibility.
```

>`--learning_rate 5e-6 `

```
[INFO|2026-09-22 19:58:38] logging.py:144 >> {'loss': 1.5199, 'learning_rate': 4.9995e-05, 'epoch': 0.12, 'throughput': 466.97}
[INFO|2026-09-22 19:58:53] logging.py:144 >> {'loss': 1.5383, 'learning_rate': 4.9975e-05, 'epoch': 0.16, 'throughput': 457.12}
[INFO|2026-09-22 19:59:08] logging.py:144 >> {'loss': 1.5237, 'learning_rate': 4.9926e-05, 'epoch': 0.20, 'throughput': 443.25}
[INFO|2026-09-22 19:59:23] logging.py:144 >> {'loss': 1.6987, 'learning_rate': 4.9852e-05, 'epoch': 0.24, 'throughput': 432.77}
[INFO|2026-09-22 19:59:38] logging.py:144 >> {'loss': 1.4719, 'learning_rate': 4.9752e-05, 'epoch': 0.28, 'throughput': 445.15}
[INFO|2026-09-22 19:59:53] logging.py:144 >> {'loss': 1.6459, 'learning_rate': 4.9626e-05, 'epoch': 0.32, 'throughput': 447.73}
[INFO|2026-09-22 20:00:08] logging.py:144 >> {'loss': 1.5147, 'learning_rate': 4.9476e-05, 'epoch': 0.36, 'throughput': 446.68}
[INFO|2026-09-22 20:00:22] logging.py:144 >> {'loss': 1.5923, 'learning_rate': 4.9300e-05, 'epoch': 0.40, 'throughput': 453.17}
[INFO|2026-09-22 20:00:37] logging.py:144 >> {'loss': 1.5026, 'learning_rate': 4.9099e-05, 'epoch': 0.44, 'throughput': 455.64}
[INFO|2026-09-22 20:00:52] logging.py:144 >> {'loss': 1.6360, 'learning_rate': 4.8874e-05, 'epoch': 0.48, 'throughput': 453.37}
[INFO|2026-09-22 20:01:07] logging.py:144 >> {'loss': 1.6738, 'learning_rate': 4.8623e-05, 'epoch': 0.52, 'throughput': 450.80}
[INFO|2026-09-22 20:01:22] logging.py:144 >> {'loss': 1.5389, 'learning_rate': 4.8349e-05, 'epoch': 0.56, 'throughput': 460.29}
[INFO|2026-09-22 20:01:37] logging.py:144 >> {'loss': 1.6136, 'learning_rate': 4.8051e-05, 'epoch': 0.60, 'throughput': 465.38}
[INFO|2026-09-22 20:01:51] logging.py:144 >> {'loss': 1.5274, 'learning_rate': 4.7729e-05, 'epoch': 0.64, 'throughput': 464.82}
[INFO|2026-09-22 20:02:06] logging.py:144 >> {'loss': 1.5205, 'learning_rate': 4.7383e-05, 'epoch': 0.68, 'throughput': 475.79}
[INFO|2026-09-22 20:02:20] logging.py:144 >> {'loss': 1.6452, 'learning_rate': 4.7015e-05, 'epoch': 0.72, 'throughput': 471.83}
[INFO|2026-09-22 20:02:35] logging.py:144 >> {'loss': 1.5653, 'learning_rate': 4.6624e-05, 'epoch': 0.76, 'throughput': 472.36}
[INFO|2026-09-22 20:02:50] logging.py:144 >> {'loss': 1.7384, 'learning_rate': 4.6211e-05, 'epoch': 0.80, 'throughput': 474.34}
[INFO|2026-09-22 20:03:04] logging.py:144 >> {'loss': 1.6410, 'learning_rate': 4.5776e-05, 'epoch': 0.84, 'throughput': 475.46}
[INFO|2026-09-22 20:03:19] logging.py:144 >> {'loss': 1.7369, 'learning_rate': 4.5319e-05, 'epoch': 0.88, 'throughput': 474.55}
[INFO|2026-09-22 20:03:34] logging.py:144 >> {'loss': 1.7003, 'learning_rate': 4.4842e-05, 'epoch': 0.92, 'throughput': 473.63}
[INFO|2026-09-22 20:03:48] logging.py:144 >> {'loss': 1.6998, 'learning_rate': 4.4344e-05, 'epoch': 0.96, 'throughput': 472.79}
[INFO|2026-09-22 20:04:03] logging.py:144 >> {'loss': 1.5017, 'learning_rate': 4.3932e-05, 'epoch': 1.00, 'throughput': 473.42}
[INFO|2026-09-22 20:04:18] logging.py:144 >> {'loss': 0.7905, 'learning_rate': 4.3399e-05, 'epoch': 1.04, 'throughput': 473.02}
[INFO|2026-09-22 20:04:33] logging.py:144 >> {'loss': 0.7134, 'learning_rate': 4.2847e-05, 'epoch': 1.08, 'throughput': 471.88}
[INFO|2026-09-22 20:04:47] logging.py:144 >> {'loss': 0.7155, 'learning_rate': 4.2277e-05, 'epoch': 1.12, 'throughput': 473.45}
[INFO|2026-09-22 20:05:02] logging.py:144 >> {'loss': 0.7559, 'learning_rate': 4.1689e-05, 'epoch': 1.16, 'throughput': 468.92}
[INFO|2026-09-22 20:05:17] logging.py:144 >> {'loss': 0.6375, 'learning_rate': 4.1083e-05, 'epoch': 1.20, 'throughput': 468.73}
[INFO|2026-09-22 20:05:31] logging.py:144 >> {'loss': 0.7212, 'learning_rate': 4.0462e-05, 'epoch': 1.24, 'throughput': 469.10}
[INFO|2026-09-22 20:05:46] logging.py:144 >> {'loss': 0.6469, 'learning_rate': 3.9824e-05, 'epoch': 1.28, 'throughput': 471.38}
[INFO|2026-09-22 20:06:01] logging.py:144 >> {'loss': 0.7402, 'learning_rate': 3.9171e-05, 'epoch': 1.32, 'throughput': 469.79}
[INFO|2026-09-22 20:06:15] logging.py:144 >> {'loss': 0.7562, 'learning_rate': 3.8504e-05, 'epoch': 1.36, 'throughput': 470.13}
[INFO|2026-09-22 20:06:30] logging.py:144 >> {'loss': 0.7701, 'learning_rate': 3.7822e-05, 'epoch': 1.40, 'throughput': 469.77}
[INFO|2026-09-22 20:06:45] logging.py:144 >> {'loss': 0.6770, 'learning_rate': 3.7128e-05, 'epoch': 1.44, 'throughput': 469.75}
[INFO|2026-09-22 20:06:59] logging.py:144 >> {'loss': 0.7002, 'learning_rate': 3.6421e-05, 'epoch': 1.48, 'throughput': 471.92}
[INFO|2026-09-22 20:07:14] logging.py:144 >> {'loss': 0.6480, 'learning_rate': 3.5702e-05, 'epoch': 1.52, 'throughput': 470.95}
[INFO|2026-09-22 20:07:29] logging.py:144 >> {'loss': 0.6811, 'learning_rate': 3.4973e-05, 'epoch': 1.56, 'throughput': 471.00}
[INFO|2026-09-22 20:07:44] logging.py:144 >> {'loss': 0.7414, 'learning_rate': 3.4233e-05, 'epoch': 1.60, 'throughput': 471.83}
[INFO|2026-09-22 20:07:58] logging.py:144 >> {'loss': 0.6542, 'learning_rate': 3.3484e-05, 'epoch': 1.64, 'throughput': 470.79}
[INFO|2026-09-22 20:08:11] logging.py:144 >> {'loss': 0.6698, 'learning_rate': 3.2725e-05, 'epoch': 1.68, 'throughput': 472.63}
[INFO|2026-09-22 20:08:24] logging.py:144 >> {'loss': 0.6626, 'learning_rate': 3.1959e-05, 'epoch': 1.72, 'throughput': 473.70}
[INFO|2026-09-22 20:08:39] logging.py:144 >> {'loss': 0.7017, 'learning_rate': 3.1186e-05, 'epoch': 1.76, 'throughput': 474.80}
[INFO|2026-09-22 20:08:54] logging.py:144 >> {'loss': 0.7056, 'learning_rate': 3.0407e-05, 'epoch': 1.80, 'throughput': 473.56}
[INFO|2026-09-22 20:09:08] logging.py:144 >> {'loss': 0.7699, 'learning_rate': 2.9622e-05, 'epoch': 1.84, 'throughput': 472.58}
[INFO|2026-09-22 20:09:23] logging.py:144 >> {'loss': 0.6723, 'learning_rate': 2.8832e-05, 'epoch': 1.88, 'throughput': 473.51}
[INFO|2026-09-22 20:09:38] logging.py:144 >> {'loss': 0.6973, 'learning_rate': 2.8038e-05, 'epoch': 1.92, 'throughput': 476.06}
[INFO|2026-09-22 20:09:53] logging.py:144 >> {'loss': 0.7110, 'learning_rate': 2.7241e-05, 'epoch': 1.96, 'throughput': 474.35}
[INFO|2026-09-22 20:10:08] logging.py:144 >> {'loss': 0.6331, 'learning_rate': 2.6442e-05, 'epoch': 2.00, 'throughput': 476.37}
[INFO|2026-09-22 20:10:22] logging.py:144 >> {'loss': 0.2584, 'learning_rate': 2.5641e-05, 'epoch': 2.04, 'throughput': 476.17}
[INFO|2026-09-22 20:10:37] logging.py:144 >> {'loss': 0.2110, 'learning_rate': 2.4840e-05, 'epoch': 2.08, 'throughput': 475.93}
```

**Comment by hsa-online** [comment]
I checked new version on the same Qwen2.5-0.5B model.
And now I'm getting the same loss at the end of training both with and without FA2.
Thank you very much!
There are some plans to continue tests on larger Qwen models, but this problem seems to be solved.
